bpo-30931: Fix asyncore race condition

vstinner opened the pull request:
poll() and poll2() now get objects from the asyncore map before
calling handlers, since handlers can modify the map.

Detect also if a ready dispatcher was closed by a previous handler,
before its handler is executed. Don't run handlers of closed
dispatchers, even if the file descriptor was reused by a newly
registered dispatcher.

Co-Authored-By: Jaume Marhuenda <jaume.marhuenda@datastax.com>

<!-- issue-number: bpo-30931 -->
https://bugs.python.org/issue30931
<!-- /issue-number -->

vstinner committed:
bpo-30931: Fix asyncore race condition

poll() and poll2() now get objects from the asyncore map before
calling handlers, since handlers can modify the map.

Detect also if a ready dispatcher was closed by a previous handler,
before its handler is executed. Don't run handlers of closed
dispatchers, even if the file descriptor was reused by a newly
registered dispatcher.

Co-Authored-By: Jaume Marhuenda <jaume.marhuenda@datastax.com>

vstinner commented:
cc @bjmb and @nirs

nirs commented on Lib/asyncore.py:
```diff
@@ -143,22 +143,17 @@ def poll(timeout=0.0, map=None):
 
-        for fd in w:
-            obj = map.get(fd)
-            if obj is None:
```
This was detecting (sometimes) closed dispatchers...

nirs commented on Lib/asyncore.py:
```diff
@@ -143,22 +143,17 @@ def poll(timeout=0.0, map=None):
-            if obj is None:
-                continue
+        for obj in robjs:
+            read(obj)
```
And now we try to read from close dispatcher.

I think we should wait for https://github.com/python/cpython/pull/2804, and do:

    for obj in ...:
        if obj.closing:
            continue
        read(obj)

And same for write/_exception.

nirs commented on Lib/asyncore.py:
```diff
@@ -181,10 +176,12 @@ def poll2(timeout=0.0, map=None):
+        # since handlers can modify map
+        ready = [(map[fd], flags) for fd, flags in r]
+
+        for obj, flags in ready:
```
Same, need to avoid calling closed dispatchers.

vstinner commented:
> Co-Authored-By: bjmb jaume.marhuenda@datastax.com

@bjmb: Would you mind to give me your full name to add you to Misc/ACKS?

nirs commented on Lib/test/test_asyncore.py:
```diff
@@ -843,5 +845,49 @@ class TestAPI_UseUnixSocketsSelect(TestAPI_UseUnixSockets, unittest.TestCase):
+                return False
+            def handle_read_event(self):
+                read_calls.append(self.fd)
+                testmap.clear()
```
This does not simulate the bug :
1. closing dispatcher in handle_read
2. creating new one, get same fd
3. asyncore.poll/poll2 access new dispatchers by mistake

Does it fail without the fix?

vstinner replied:
> Does it fail without the fix?

Yes, I checked.

vstinner replied:
In the meanwhile, I added a new test_poll_close_replace_dispatchers() test which tests exactly the described behaviour.

beltran commented:
@haypo my full name is Jaume Marhuenda,  about this PR I think there could be a bug if a socket in `robjs` closed a socket in `wobjs`, we would still call `write(obj)` and then `handle_write` for that socket even if closed

vstinner commented:
@bjmb: "my full name is Jaume Marhuenda"

Ok, I added you to Misc/ACKS and completed your name in my commit message.

nirs commented:
@giampaolo can you take a look?

vstinner commented:
@bjmb: Would you mind to review my PR? Would you be ok to abandon your PR 2707 in favor of my PR? See https://bugs.python.org/issue30931#msg299164 for my rationale.

giampaolo commented on Lib/asyncore.py:
```diff
@@ -143,21 +143,25 @@ def poll(timeout=0.0, map=None):
+            if map.get(fd) is not obj:
+                # bpo-30931: obj has been closed by a previously executed
+                # handler.  Its file descriptor may have been reused by new
+                # dispatcher registered later.
```
To my understanding an instance with the same fd can be replaced only after `read()` occurs so retrieving all `robjs` upfront should not be necessary.

nirs replied:
You must fetch all the instances upfront, since a read may replace any other dispatcher referenced in robjs, wobjs, and eobjs lists.

giampaolo replied:
Oh right. Since the original check was `if obj is None:` shouldn't this be `if map.get(fd) is not obj or obj is None:`?

vstinner replied:
> Since the original check was if obj is None: shouldn't this be if map.get(fd) is not obj or obj is None:?

obj cannot be None.

The current code checks if fd is still in map because map can be modified while we iterate on ready file descriptors. My PR doesn't have this specific issue, since it iterates on a separated list.

vstinner committed:
Fix another poll2() race condition

If a single fd requires to call multiple handlers (ex: readable and
writable dispatcher) and a handler closes the dispatcher, don't call
following handlers.

Inline readwrite() and check if the dispatcher was closed before
calling each handler.

nirs commented on Lib/asyncore.py:
```diff
@@ -191,12 +191,40 @@ def poll2(timeout=0.0, map=None):
+            except _reraised_exceptions:
+                raise
+            except:
+                obj.handle_error()
```
Not sure about this, I would replace select to work with readwrite instead, and solve the issue in readwrite or in the handle_xxx_events.

For example, we can do:

    def handle_read_event(self):
        if self._fileno is None:
            raise DispatcherClosed

And in readwrite:

    try:
        call handlers ...
    except DispatcherClosed:
        ...
    
Note also that some handlers call other handlers, so this fix is not enough:

```python
    def handle_read_event(self):
        if self.accepting:
            # accepting sockets are never connected, they "spawn" new
            # sockets that are connected
            self.handle_accept()
        elif not self.connected:
            if self.connecting:
                self.handle_connect_event() 
            self.handle_read()
        else:
            self.handle_read()
```

If dispatcher was closed in handle_connect_event(), we will call handle_read() :-)

vstinner replied:
Oh sorry, I misunderstood your proposal. I was confused between handle_read_event() defined in Lib/asyncore.py and handle_read() which is user defined.

> If dispatcher was closed in handle_connect_event(), we will call handle_read() :-)

Right. Oh god, I didn't expect such many race conditions.

```
def handle_read_event(self):
    if self._fileno is None:
        raise DispatcherClosed
```

Why not just "return", do nothing if the dispatcher was closed? Existing example:

```
    def handle_write_event(self):
        if self.accepting:
            # Accepting sockets shouldn't get a write event.
            # We will pretend it didn't happen.
            return

        if not self.connected:
            if self.connecting:
                self.handle_connect_event()
        self.handle_write()
```

vstinner replied:
Please see my long summary at http://bugs.python.org/issue30931#msg299300

IMHO fixing handle_write_events() deserves its own bpo: http://bugs.python.org/issue30985

I suggest to first merge this PR, and later discussed what I called the "(Bug 4)".

vstinner committed:
Merge branch 'master' into asyncore_ready

vstinner committed:
Rewrite the NEWS entry.

vstinner commented:
I rewrote the NEWS entry to better describe the **behaviour change**, rather than focusing on the implementation change.

@nirs, @giampaolo: would you mind to review the latest version of this PR?

See also http://bugs.python.org/issue30931#msg299300 for my long  "summary" of all bugs we discussed and my proposed plan to fix them.

csabella commented on Misc/NEWS.d/next/Library/2017-07-24-23-52-39.bpo-30931.mZMsJ7.rst:
```diff
@@ -0,0 +1,3 @@
+asyncore: poll() and poll2() functions now check if a dispatcher was closed (if
+it's file descriptor is no more in the map) before calling its handlers:
+handlers of closed dispatchers are no more called.
```
if it's file descriptor -> if its file descriptor
is no more in the map -> is no longer in the map?
no more called -> no longer called

nirs replied:
poll() and poll2() already check if a dispatcher was closed before calling it. The change in behavior is detecting a replaced dispatcher using same fd, and skipping it instead of calling it before it is ready.

vstinner replied:
@nirs: Oops, you are right. I rewrote the NEWS entry.

@csabella: fixed, thanks.

vstinner committed:
Fix the NEWS entry

asyncore already detected if a dispatcher was closed.

vstinner committed:
Merge branch 'master' into asyncore_ready

vstinner committed:
Rename the unit test: test_poll_close_reuse_fd()

vstinner committed:
NEWS: no more => no longer
